### DISCUSSION TITLE
how to get hls stream?

### DISCUSSION THREAD

**Discussion by wbtcpip2** [bluenviron/mediamtx#1675] [Questions]
hi, i jyst started using MediaMTX on windows 10.

my encoder is sending an mpeg-ts stream with h264 video and aac audio codecs via rtsp.

i can play the stream with vlc and other players with rtsp://ip:8554/live

accoding to the features of MediaMTX Streams are automatically converted from a protocol to another. For instance, it's possible to publish a stream with RTSP and read it with HLS

but how i can read my stream with HLS now?

i tried http://ip:8888/live/index.m3u8 with vlc but i get

2023/04/10 21:55:56 INF [HLS] [muxer live] created (requested by xxx.xxx.xxx.xxx)
2023/04/10 21:55:56 INF [HLS] [muxer live] destroyed (the stream doesn't contain any supported codec (which are currently H264, H265, MPEG4-Audio, Opus))

what i have to do in order to play the HLS stream?

thank you

**Comment by dams92210** [comment]
Use stream.m3u8 instead of index.m3u8.

**Comment by wbtcpip2** [comment]
unfortunately still the same.

when i start publishing with my rtsp encoder (gstreamer) i see:

2023/04/11 13:24:41 INF [RTSP] [conn 127.0.0.1:60148] opened
2023/04/11 13:24:42 INF [RTSP] [session c3c47d0a] created by 127.0.0.1:60148
2023/04/11 13:24:43 INF [RTSP] [session c3c47d0a] is publishing to path 'live', with UDP, 1 track (Generic)

when i read rtsp://192.168.1.201:8554/live it works and i see

2023/04/11 15:41:07 INF [RTSP] [conn 192.168.1.202:56969] opened
2023/04/11 15:41:07 INF [RTSP] [session 3f92ec17] created by 192.168.1.202:56969
2023/04/11 15:41:07 INF [RTSP] [session 3f92ec17] is reading from path 'live', with UDP, 1 track (Generic)

when i read http://192.168.1.201:8888/live/stream.m3u8 i see:

2023/04/11 15:42:41 INF [HLS] [muxer live] created (requested by 192.168.1.202)
2023/04/11 15:42:41 INF [HLS] [muxer live] destroyed (the stream doesn't contain any supported codec (which are currently H264, H265, MPEG4-Audio, Opus))

**Comment by wbtcpip2** [comment]
[mediamtx.log](https://github.com/aler9/mediamtx/files/11202603/mediamtx.log)

**Reply by dams92210** [reply]
Have you tried sending with ffmpeg? The issue might be related to they way you are sending through gstreamer.

**Comment by wbtcpip2** [comment]
you are right!. gstreamer rtspclientsink is incompatible with mediamtx. changed gstreamer to rtmpsink works now.

**Comment by aler9** [comment]
Hello, there are a lot of imprecise informations here:

* rtspclientsink is fully compatible with the server. Your issue is that you're trying to convert an MPEG-TS-through-RTSP stream into HLS, which is unsupported, as clearly stated in the log messag. In order to convert an H264 and an AAC track to HLS, you have to publish them directly with RTSP, without encoding them into MPEG-TS. If you're using GStreamer, this means removing the mpegtsmux element.
* changing the pipeline to rtmpsink results in a change of protocol, another thing which is not recommended. RTSP supports a wide range of codecs and is faster, while RTMP is fit for legacy applications.
* index.m3u8 is the HLS multivariant playlist, while stream.m3u8 is the HLS media playlist. Per specification, a client must point to the multivariant playlist, that is the former.

My advice is to take a look at the README which already contains instructions on how to publish from GStreamer and read from HLS.

**Comment by wbtcpip2** [comment]
thank you for your reply.
how i can mux audio and video without using mpegtsmux in gstreamer?
looking at this exemple pipeline how i must change it?

gst-launch-1.0.exe -e videotestsrc ! x264enc bitrate=1856 ! video/x-h264 ! h264parse ! queue ! mpegtsmux name=mux ! rtspclientsink location=rtsp://192.168.1.201:8554/live audiotestsrc ! audioconvert ! audioresample ! audio/x-raw, rate=48000 ! mfaacenc bitrate=192000 ! aacparse ! mux.

because in mediamtx produce an unknown track

2023/04/12 11:38:19 INF [RTSP] [session 4c7db001] is publishing to path 'live', with UDP, 1 track (Generic)
2023/04/12 11:38:19 DEB [RTSP] [conn 192.168.1.202:49544] [s->c] RTSP/1.0 200 OK

and the mediamtx hls server do not work with this.

so how i can mux video+audio without mpegtsmux in gstreamer?

**Comment by wbtcpip2** [comment]
ok now this pipeline works:

gst-launch-1.0.exe -e videotestsrc ! x264enc bitrate=1856 ! video/x-h264 ! queue ! rtspclientsink location=rtsp://192.168.1.201:8554/live

2023/04/12 11:43:00 INF [RTSP] [session cde9c8b7] is publishing to path 'live', with UDP, 1 track (H264)
2023/04/12 11:43:00 DEB [RTSP] [conn 192.168.1.202:49562] [s->c] RTSP/1.0 200 OK

but i do not have audio

**Comment by wbtcpip2** [comment]
solved with this pipeline

gst-launch-1.0.exe -e videotestsrc ! mfh264enc bitrate=1024 ! video/x-h264 ! rtspclientsink name=mux location=rtsp://192.168.1.201:8554/live audiotestsrc ! audioconvert ! audioresample ! audio/x-raw, rate=48000 ! mfaacenc bitrate=192000 ! audio/mpeg, mpegversion=4, stream-format=raw ! mux.

thank you

[a comment by an automated account is not shown]
